src/target/riscv/riscv-013.c

```diff
@@ -2426,130 +2426,95 @@ static int init_target(struct command_context *cmd_ctx,
+	if (result != ERROR_OK)
+		return result;
 
 	uint32_t dmstatus;
-	int dmi_busy_delay = info->dmi_busy_delay;
+	const int dmi_busy_delay = info->dmi_busy_delay;
 	time_t start = time(NULL);
-
-	for (unsigned int i = 0; i < riscv_count_harts(target); ++i) {
-		unsigned int index = i;
-		if (target->rtos) {
-			if (index != info->index)
-				continue;
-			dmi_write(target, DM_DMCONTROL,
-					set_dmcontrol_hartsel(control_haltreq, index));
-		} else {
-			index = info->index;
-		}
-
-		LOG_DEBUG("Waiting for hart %d to come out of reset.", index);
-		while (1) {
-			int result = dmstatus_read_timeout(target, &dmstatus, true,
+	LOG_TARGET_DEBUG(target, "Waiting for hart to come out of reset.");
+	do {
+		result = dmstatus_read_timeout(target, &dmstatus, true,
+				riscv_reset_timeout_sec);
+		if (result == ERROR_TIMEOUT_REACHED)
+			LOG_TARGET_ERROR(target, "Hart didn't complete a DMI read coming "
+					"out of reset in %ds; Increase the timeout with riscv "
+					"set_reset_timeout_sec.",
 					riscv_reset_timeout_sec);
-			if (result == ERROR_TIMEOUT_REACHED)
-				LOG_ERROR("Hart %d didn't complete a DMI read coming out of "
-						"reset in %ds; Increase the timeout with riscv "
-						"set_reset_timeout_sec.",
-						index, riscv_reset_timeout_sec);
-			if (result != ERROR_OK)
-				return result;
-			/* Certain debug modules, like the one in GD32VF103
-			 * MCUs, violate the specification's requirement that
-			 * each hart is in "exactly one of four states" and,
-			 * during reset, report harts as both unavailable and
-			 * halted/running. To work around this, we check for
-			 * the absence of the unavailable state rather than
-			 * the presence of any other state. */
-			if (!get_field(dmstatus, DM_DMSTATUS_ALLUNAVAIL))
-				break;
-			if (time(NULL) - start > riscv_reset_timeout_sec) {
-				LOG_ERROR("Hart %d didn't leave reset in %ds; "
-						"dmstatus=0x%x; "
-						"Increase the timeout with riscv set_reset_timeout_sec.",
-						index, riscv_reset_timeout_sec, dmstatus);
-				return ERROR_FAIL;
-			}
-		}
-		if (target->reset_halt) {
-			target->state = TARGET_HALTED;
-			target->debug_reason = DBG_REASON_DBGRQ;
-		} else {
-			target->state = TARGET_RUNNING;
-			target->debug_reason = DBG_REASON_NOTHALTED;
-		}
+		if (result != ERROR_OK)
+			return result;
 
-		if (get_field(dmstatus, DM_DMSTATUS_ALLHAVERESET)) {
-			/* Ack reset and clear DM_DMCONTROL_HALTREQ if previously set */
-			dmi_write(target, DM_DMCONTROL,
-					set_dmcontrol_hartsel(control, index) |
-					DM_DMCONTROL_ACKHAVERESET);
+		if (time(NULL) - start > riscv_reset_timeout_sec) {
+			LOG_TARGET_ERROR(target, "Hart didn't leave reset in %ds; "
+					"dmstatus=0x%x (allunavail=%s, allhavereset=%s); "
+					"Increase the timeout with riscv set_reset_timeout_sec.",
+					riscv_reset_timeout_sec, dmstatus,
+					get_field(dmstatus, DM_DMSTATUS_ALLUNAVAIL) ? "true" : "false",
+					get_field(dmstatus, DM_DMSTATUS_ALLUNAVAIL) ? "true" : "false");
```

Comment:
There seems to be a copy-paste typo:

```suggestion
					get_field(dmstatus, DM_DMSTATUS_ALLHAVERESET) ? "true" : "false");
```

Reply:
Addressed.